client/gtk/src/main.rs

```diff
@@ -44,32 +44,75 @@ impl Win {
     fn handle_action(&mut self, action: Action) -> () {
         match action {
             Action::AddMessage(msg) => {
-                let text_buffer = self.widgets.messages.get_buffer().unwrap();
-                let message = format!("{}: {}\n", msg.author, msg.content);
-                text_buffer.insert(&mut text_buffer.get_end_iter(), &message);
+                self.push_message(&msg.author, &msg.content);
+            }
+            Action::AddRoom(room) => {
+                self.push_message("Info", &format!("Joined room {}", room.0));
+
+                self.model.room = Some(room);
+                let txt: &str = &format!("#{}", room.0);
+                let room_label = Label::new(Some(txt));
+                self.widgets.rooms.insert(&room_label, -1);
+                self.model.room_list.push(room);
+                room_label.show_all();
             }
```

Comment:
Is there some kind of callback message for a successful login? Or otherwise how does the GUI know whether it was successful or not?

Reply:
There isn't, but there is an error sent if login is unsuccessful

Reply:
Maybe in future this will be required. Watch this space...